### ISSUE TITLE
Ansible stderr often gets swallowed

### ISSUE THREAD

**Issue by greg-hellings** [RedHatQE/teflo#54]
When runs of ansible-playbook error, they often result in none of their own stderr being output. Instead, I typically get a Python stack trace from inside of teflo. This then requires me to run the playbook again, manually, in order to see the output error.

Please don't swallow stderr

**Comment by rujutashinde** [comment]
https://projects.engineering.redhat.com/browse/CCITCARBON-281 filed for this

**Comment by JunqiZhang0** [comment]
Scheduled a meeting to discuss this issue

**Comment by JunqiZhang0** [comment]
Had the meeting with Greg, @greg-hellings will have a minimum ansible playbook attached

**Comment by greg-hellings** [comment]
@JunqiZhang0 Making the min reproducer was easier than I thought. Here are my sdf.yml, min.yml (the playbook), along with the output from Teflo and the output from `ansible-playbook` that is mostly getting swallowed by Teflo:

https://gist.github.com/greg-hellings/8b14a434302055d077ef69ab5360fe4a

**Comment by jbpratt** [comment]
@JunqiZhang0 @rujutashinde  I'm still experiencing this issue on `teflo, version 1.2.0`

`.teflo/fc0vlz8i25/logs/teflo_scenario.log`

<details>

```
2021-05-18 07:50:33,037 WARNING Scenario workspace was not set, therefore the workspace is automatically assigned to the current working directory. You may experience problems if files needed by teflo do not exists in the scenario workspace.
2021-05-18 07:50:33,221 INFO 

2021-05-18 07:50:33,221 INFO                                TEFLO RUN (START)                               
2021-05-18 07:50:33,221 INFO -------------------------------------------------------------------------------
2021-05-18 07:50:33,222 INFO  * Data Folder           : .teflo/fc0vlz8i25
2021-05-18 07:50:33,222 INFO  * Workspace             : /home/bpratt/pit/scenarios/sbo_openshift/venv/sbo_openshift
2021-05-18 07:50:33,222 INFO  * Log Level             : info
2021-05-18 07:50:33,223 INFO  * Tasks                 : ['orchestrate']
2021-05-18 07:50:33,223 INFO  * Scenario              : Layered product OpenShift unit test scenario
2021-05-18 07:50:33,223 INFO  * Included Scenario(s)  : ['resource_check', 'provision_base', 'orchestrate_base', 'orchestrate_install_openshift_openstack', 'orchestrate', 'execute']
2021-05-18 07:50:33,224 INFO -------------------------------------------------------------------------------

2021-05-18 07:50:33,224 INFO  * Task    : orchestrate
2021-05-18 07:50:33,224 INFO Sending out any notifications that are registered.
2021-05-18 07:50:33,225 INFO ..................................................
2021-05-18 07:50:33,225 INFO Starting tasks on pipeline: notify
2021-05-18 07:50:33,226 WARNING ... no tasks to be executed ...
2021-05-18 07:50:33,226 INFO ..................................................
2021-05-18 07:50:33,227 INFO Starting tasks on pipeline: orchestrate
2021-05-18 07:50:33,227 INFO --> Blaster v0.4.0 <--
2021-05-18 07:50:33,227 INFO Task Execution: Sequential
2021-05-18 07:50:33,228 INFO Tasks:
2021-05-18 07:50:33,228 INFO 1. Task     : Move interop-qe-openshift files to scenario's ansible directory
                                Class    : <class 'teflo.tasks.orchestrate.OrchestrateTask'>
                                Methods  : ['run']
2021-05-18 07:50:33,229 INFO 2. Task     : Configure OpenShift test driver
                                Class    : <class 'teflo.tasks.orchestrate.OrchestrateTask'>
                                Methods  : ['run']
2021-05-18 07:50:33,229 INFO 3. Task     : Download OpenShift client 'oc'
                                Class    : <class 'teflo.tasks.orchestrate.OrchestrateTask'>
                                Methods  : ['run']
2021-05-18 07:50:33,229 INFO 4. Task     : Configure environment for administrator access to OCP cluster
                                Class    : <class 'teflo.tasks.orchestrate.OrchestrateTask'>
                                Methods  : ['run']
2021-05-18 07:50:33,230 INFO 5. Task     : Create OpenShift cluster in OpenStack
                                Class    : <class 'teflo.tasks.orchestrate.OrchestrateTask'>
                                Methods  : ['run']
2021-05-18 07:50:33,230 INFO 6. Task     : Setup for running Service Binding Operator tests
                                Class    : <class 'teflo.tasks.orchestrate.OrchestrateTask'>
                                Methods  : ['run']
2021-05-18 07:50:33,230 INFO ** BLASTER BEGIN **
2021-05-18 07:50:33,232 INFO    running orchestration Move interop-qe-openshift files to scenario's ansible directory for ['localhost']
2021-05-18 07:50:33,236 INFO Executing playbook:
2021-05-18 07:50:33,236 INFO Ansible options used: {}
2021-05-18 07:50:33,236 INFO Executing playbook : ansible/relocate_files.yml
2021-05-18 07:50:33,651 INFO ansible-playbook 2.10.9
2021-05-18 07:50:33,652 INFO config file = /home/bpratt/pit/scenarios/sbo_openshift/venv/sbo_openshift/ansible.cfg
2021-05-18 07:50:33,652 INFO configured module search path = ['/home/bpratt/pit/scenarios/sbo_openshift/venv/sbo_openshift/ansible/interop_qe_openshift/ansible/library']
2021-05-18 07:50:33,652 INFO ansible python module location = /home/bpratt/pit/scenarios/sbo_openshift/venv/lib64/python3.9/site-packages/ansible
2021-05-18 07:50:33,652 INFO executable location = /home/bpratt/pit/scenarios/sbo_openshift/venv/bin/ansible-playbook
2021-05-18 07:50:33,652 INFO python version = 3.9.4 (default, Apr  6 2021, 00:00:00) [GCC 11.0.1 20210324 (Red Hat 11.0.1-0)]
2021-05-18 07:50:33,652 INFO Using /home/bpratt/pit/scenarios/sbo_openshift/venv/sbo_openshift/ansible.cfg as config file
2021-05-18 07:50:33,743 INFO redirecting (type: callback) ansible.builtin.yaml to community.general.yaml
2021-05-18 07:50:33,752 INFO redirecting (type: callback) ansible.builtin.yaml to community.general.yaml
2021-05-18 07:50:33,760 INFO Skipping callback 'default', as we already have a stdout callback.
2021-05-18 07:50:33,760 INFO Skipping callback 'minimal', as we already have a stdout callback.
2021-05-18 07:50:33,760 INFO Skipping callback 'oneline', as we already have a stdout callback.
2021-05-18 07:50:33,760 INFO 
2021-05-18 07:50:33,760 INFO PLAYBOOK: relocate_files.yml ***************************************************
2021-05-18 07:50:33,760 INFO 1 plays in ansible/relocate_files.yml
2021-05-18 07:50:33,762 INFO 
2021-05-18 07:50:33,762 INFO PLAY [Relocate Interop QE OpenShift automation files for easier scenario access] ***
2021-05-18 07:50:33,773 INFO 
2021-05-18 07:50:33,773 INFO TASK [Gathering Facts] *********************************************************
2021-05-18 07:50:33,773 INFO task path: /home/bpratt/pit/scenarios/sbo_openshift/venv/sbo_openshift/ansible/relocate_files.yml:2
2021-05-18 07:50:34,596 INFO ok: [localhost]
2021-05-18 07:50:34,604 INFO META: ran handlers
2021-05-18 07:50:34,611 INFO 
2021-05-18 07:50:34,612 INFO TASK [Copy OpenShift automation files to scenarios ansible directory] **********
2021-05-18 07:50:34,612 INFO task path: /home/bpratt/pit/scenarios/sbo_openshift/venv/sbo_openshift/ansible/relocate_files.yml:14
2021-05-18 07:50:50,272 INFO ok: [localhost] => changed=false
2021-05-18 07:50:50,272 INFO dest: ./
2021-05-18 07:50:50,273 INFO src: /home/bpratt/pit/scenarios/sbo_openshift/venv/interop_qe_openshift
2021-05-18 07:50:50,279 INFO META: ran handlers
2021-05-18 07:50:50,286 INFO META: ran handlers
2021-05-18 07:50:50,287 INFO 
2021-05-18 07:50:50,287 INFO PLAY RECAP *********************************************************************
2021-05-18 07:50:50,287 INFO localhost                  : ok=2    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
2021-05-18 07:50:50,287 INFO 
2021-05-18 07:50:50,344 INFO Successfully completed playbook : ansible/relocate_files.yml
2021-05-18 07:50:50,408 INFO Orchestration passed : Successfully completed orchestrate action: Move interop-qe-openshift files to scenario's ansible directory.
2021-05-18 07:50:50,410 INFO    running orchestration Configure OpenShift test driver for ['ocp_driver_bpratt']
2021-05-18 07:50:50,414 INFO Installing roles using req. file: /home/bpratt/pit/scenarios/sbo_openshift/venv/sbo_openshift/ansible/interop_qe_openshift/ansible/requirements.yml
2021-05-18 07:50:50,823 INFO Starting galaxy role install process
2021-05-18 07:50:50,825 INFO - ocp_cloud_dns is already installed, skipping.
2021-05-18 07:50:50,826 INFO - install-certs is already installed, skipping.
2021-05-18 07:50:50,870 INFO Roles installed successfully from: /home/bpratt/pit/scenarios/sbo_openshift/venv/sbo_openshift/ansible/interop_qe_openshift/ansible/requirements.yml!
2021-05-18 07:50:50,872 INFO Executing playbook:
2021-05-18 07:50:50,872 INFO Ansible options used: {}
2021-05-18 07:50:50,872 INFO Executing playbook : ansible/interop_qe_openshift/ansible/playbooks/setup_test_driver.yml
2021-05-18 07:50:50,877 INFO x.x.x.x
2021-05-18 07:50:50,878 ERROR Orchestration failed : 
2021-05-18 07:50:50,914 ERROR Failed to run orchestration Configure OpenShift test driver 
2021-05-18 07:50:50,915 ERROR Orchestration failed : Failed to perform  Configure OpenShift test driver
2021-05-18 07:50:50,916 ERROR Traceback (most recent call last):
  File "/home/bpratt/pit/scenarios/sbo_openshift/venv/lib64/python3.9/site-packages/teflo/tasks/orchestrate.py", line 59, in run
    self.orchestrator.run()
  File "/home/bpratt/pit/scenarios/sbo_openshift/venv/lib64/python3.9/site-packages/teflo/orchestrators/action_orchestrator.py", line 68, in run
    raise TefloOrchestratorError("Orchestration failed : Failed to perform  %s" % self.plugin.action_name)
teflo.exceptions.TefloOrchestratorError: Orchestration failed : Failed to perform  Configure OpenShift test driver

2021-05-18 07:50:50,917 ERROR A exception was raised while processing task: Configure OpenShift test driver method: run
2021-05-18 07:50:51,919 INFO ** BLASTER COMPLETE **
2021-05-18 07:50:51,922 INFO     -> TOTAL DURATION: 0h:0m:18s
2021-05-18 07:50:51,925 ERROR One or more tasks got a status of non zero.
2021-05-18 07:50:51,927 INFO Sending out any notifications that are registered.
2021-05-18 07:50:51,931 INFO ..................................................
2021-05-18 07:50:51,933 INFO Starting tasks on pipeline: notify
2021-05-18 07:50:51,935 WARNING ... no tasks to be executed ...
2021-05-18 07:50:52,019 INFO 

2021-05-18 07:50:52,019 INFO                                SCENARIO RUN (END)                              
2021-05-18 07:50:52,020 INFO -------------------------------------------------------------------------------
2021-05-18 07:50:52,020 INFO  * Duration                       : 0h:0m:18s
2021-05-18 07:50:52,020 INFO  * Failed Tasks                   : ['orchestrate']
2021-05-18 07:50:52,021 INFO  * Results Folder                 : .teflo/.results
2021-05-18 07:50:52,021 INFO  * Included Scenario Definition   : ['.teflo/.results/resource_check_results.yml', '.teflo/.results/provision_base_results.yml', '.teflo/.results/orchestrate_base_results.yml', '.teflo/.results/orchestrate_install_openshift_openstack_results.yml', '.teflo/.results/orchestrate_results.yml', '.teflo/.results/execute_results.yml']
2021-05-18 07:50:52,021 INFO  * Final Scenario Definition      : .teflo/.results/results.yml
2021-05-18 07:50:52,022 INFO -------------------------------------------------------------------------------
2021-05-18 07:50:52,022 INFO TEFLO RUN (RESULT=FAILED)

```

</details>


`.teflo/fc0vlz8i25/logs/ansible_orchestrator/ansible.log`
<details>

```
2021-05-18 07:47:59,266 p=212249 u=bpratt n=ansible | [WARNING]: Unable to parse /home/bpratt/pit/scenarios/sbo_openshift/venv/sbo_openshift/.teflo/.results/inventory as an inventory source

2021-05-18 07:47:59,266 p=212249 u=bpratt n=ansible | [WARNING]: No inventory was parsed, only implicit localhost is available

2021-05-18 07:47:59,685 p=212259 u=bpratt n=ansible | ansible-playbook 2.10.9
  config file = /home/bpratt/pit/scenarios/sbo_openshift/venv/sbo_openshift/ansible.cfg
  configured module search path = ['/home/bpratt/pit/scenarios/sbo_openshift/venv/sbo_openshift/ansible/interop_qe_openshift/ansible/library']
  ansible python module location = /home/bpratt/pit/scenarios/sbo_openshift/venv/lib64/python3.9/site-packages/ansible
  executable location = /home/bpratt/pit/scenarios/sbo_openshift/venv/bin/ansible-playbook
  python version = 3.9.4 (default, Apr  6 2021, 00:00:00) [GCC 11.0.1 20210324 (Red Hat 11.0.1-0)]
2021-05-18 07:47:59,685 p=212259 u=bpratt n=ansible | Using /home/bpratt/pit/scenarios/sbo_openshift/venv/sbo_openshift/ansible.cfg as config file
2021-05-18 07:47:59,686 p=212259 u=bpratt n=ansible | [WARNING]: Unable to parse /home/bpratt/pit/scenarios/sbo_openshift/venv/sbo_op
enshift/.teflo/.results/inventory as an inventory source

2021-05-18 07:47:59,686 p=212259 u=bpratt n=ansible | [WARNING]: No inventory was parsed, only implicit localhost is available

2021-05-18 07:47:59,691 p=212259 u=bpratt n=ansible | [WARNING]: provided hosts list is empty, only localhost is available. Note that
the implicit localhost does not match 'all'

2021-05-18 07:47:59,755 p=212259 u=bpratt n=ansible | redirecting (type: callback) ansible.builtin.yaml to community.general.yaml
2021-05-18 07:47:59,761 p=212259 u=bpratt n=ansible | redirecting (type: callback) ansible.builtin.yaml to community.general.yaml
2021-05-18 07:47:59,767 p=212259 u=bpratt n=ansible | Skipping callback 'default', as we already have a stdout callback.
2021-05-18 07:47:59,767 p=212259 u=bpratt n=ansible | Skipping callback 'minimal', as we already have a stdout callback.
2021-05-18 07:47:59,767 p=212259 u=bpratt n=ansible | Skipping callback 'oneline', as we already have a stdout callback.
2021-05-18 07:47:59,767 p=212259 u=bpratt n=ansible | PLAYBOOK: relocate_files.yml ***************************************************
2021-05-18 07:47:59,767 p=212259 u=bpratt n=ansible | 1 plays in ansible/relocate_files.yml
2021-05-18 07:47:59,769 p=212259 u=bpratt n=ansible | [WARNING]: Found variable using reserved name: hosts

2021-05-18 07:47:59,770 p=212259 u=bpratt n=ansible | PLAY [Relocate Interop QE OpenShift automation files for easier scenario access] ***
2021-05-18 07:47:59,778 p=212259 u=bpratt n=ansible | TASK [Gathering Facts] *********************************************************
2021-05-18 07:48:00,466 p=212259 u=bpratt n=ansible | ok: [localhost]
2021-05-18 07:48:00,475 p=212259 u=bpratt n=ansible | META: ran handlers
2021-05-18 07:48:00,484 p=212259 u=bpratt n=ansible | TASK [Copy OpenShift automation files to scenarios ansible directory] **********
2021-05-18 07:48:16,311 p=212259 u=bpratt n=ansible | ok: [localhost] => changed=false 
  dest: ./
  src: /home/bpratt/pit/scenarios/sbo_openshift/venv/interop_qe_openshift
2021-05-18 07:48:16,317 p=212259 u=bpratt n=ansible | META: ran handlers
2021-05-18 07:48:16,324 p=212259 u=bpratt n=ansible | META: ran handlers
2021-05-18 07:48:16,324 p=212259 u=bpratt n=ansible | PLAY RECAP *********************************************************************
2021-05-18 07:48:16,325 p=212259 u=bpratt n=ansible | localhost                  : ok=2    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
2021-05-18 07:50:33,651 p=212984 u=bpratt n=ansible | ansible-playbook 2.10.9
  config file = /home/bpratt/pit/scenarios/sbo_openshift/venv/sbo_openshift/ansible.cfg
  configured module search path = ['/home/bpratt/pit/scenarios/sbo_openshift/venv/sbo_openshift/ansible/interop_qe_openshift/ansible/library']
  ansible python module location = /home/bpratt/pit/scenarios/sbo_openshift/venv/lib64/python3.9/site-packages/ansible
  executable location = /home/bpratt/pit/scenarios/sbo_openshift/venv/bin/ansible-playbook
  python version = 3.9.4 (default, Apr  6 2021, 00:00:00) [GCC 11.0.1 20210324 (Red Hat 11.0.1-0)]
2021-05-18 07:50:33,651 p=212984 u=bpratt n=ansible | Using /home/bpratt/pit/scenarios/sbo_openshift/venv/sbo_openshift/ansible.cfg as config file
2021-05-18 07:50:33,743 p=212984 u=bpratt n=ansible | redirecting (type: callback) ansible.builtin.yaml to community.general.yaml
2021-05-18 07:50:33,752 p=212984 u=bpratt n=ansible | redirecting (type: callback) ansible.builtin.yaml to community.general.yaml
2021-05-18 07:50:33,759 p=212984 u=bpratt n=ansible | Skipping callback 'default', as we already have a stdout callback.
2021-05-18 07:50:33,760 p=212984 u=bpratt n=ansible | Skipping callback 'minimal', as we already have a stdout callback.
2021-05-18 07:50:33,760 p=212984 u=bpratt n=ansible | Skipping callback 'oneline', as we already have a stdout callback.
2021-05-18 07:50:33,760 p=212984 u=bpratt n=ansible | PLAYBOOK: relocate_files.yml ***************************************************
2021-05-18 07:50:33,760 p=212984 u=bpratt n=ansible | 1 plays in ansible/relocate_files.yml
2021-05-18 07:50:33,762 p=212984 u=bpratt n=ansible | [WARNING]: Found variable using reserved name: hosts

2021-05-18 07:50:33,762 p=212984 u=bpratt n=ansible | PLAY [Relocate Interop QE OpenShift automation files for easier scenario access] ***
2021-05-18 07:50:33,773 p=212984 u=bpratt n=ansible | TASK [Gathering Facts] *********************************************************
2021-05-18 07:50:34,596 p=212984 u=bpratt n=ansible | ok: [localhost]
2021-05-18 07:50:34,604 p=212984 u=bpratt n=ansible | META: ran handlers
2021-05-18 07:50:34,611 p=212984 u=bpratt n=ansible | TASK [Copy OpenShift automation files to scenarios ansible directory] **********
2021-05-18 07:50:50,272 p=212984 u=bpratt n=ansible | ok: [localhost] => changed=false 
  dest: ./
  src: /home/bpratt/pit/scenarios/sbo_openshift/venv/interop_qe_openshift
2021-05-18 07:50:50,279 p=212984 u=bpratt n=ansible | META: ran handlers
2021-05-18 07:50:50,286 p=212984 u=bpratt n=ansible | META: ran handlers
2021-05-18 07:50:50,287 p=212984 u=bpratt n=ansible | PLAY RECAP *********************************************************************
2021-05-18 07:50:50,287 p=212984 u=bpratt n=ansible | localhost                  : ok=2    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
2021-05-18 07:50:50,823 p=213600 u=bpratt n=ansible | Starting galaxy role install process
2021-05-18 07:50:50,823 p=213600 u=bpratt n=ansible | [WARNING]: - oasis_roles.ocp_install (master) is already installed - use
--force to change version to unspecified

2021-05-18 07:50:50,824 p=213600 u=bpratt n=ansible | [WARNING]: - oasis_roles.ocp_client_install (master) is already installed - use
--force to change version to unspecified

2021-05-18 07:50:50,824 p=213600 u=bpratt n=ansible | [WARNING]: - oasis_roles.index_href (master) is already installed - use --force
to change version to unspecified

2021-05-18 07:50:50,825 p=213600 u=bpratt n=ansible | [WARNING]: - oasis_roles.ocp_pull_secrets (master) is already installed - use
--force to change version to unspecified

2021-05-18 07:50:50,825 p=213600 u=bpratt n=ansible | - ocp_cloud_dns is already installed, skipping.
2021-05-18 07:50:50,826 p=213600 u=bpratt n=ansible | [WARNING]: - oasis_roles.system_repositories (master) is already installed -
use --force to change version to unspecified

2021-05-18 07:50:50,826 p=213600 u=bpratt n=ansible | - install-certs is already installed, skipping.
```

</details>


```
❯ ansible-playbook -i .teflo/.results/inventory/master-llgcbxwo70 -e @ansible/vars/sbo_v0.7.yml ansible/interop_qe_openshift/ansible/playbooks/setup_test_driver.yml

PLAY [Initialize OpenShift test driver] ******************************************************************************************************

TASK [Set python interpreter to /usr/libexec/platform-python] ********************************************************************************
ok: [xxxx]

TASK [Verify /usr/libexec/platform-python exists] ********************************************************************************************
fatal: [xxxx]: UNREACHABLE! => changed=false
  msg: |-
    Failed to connect to the host via ssh: Warning: Permanently added 'xxxx' (ED25519) to the list of known hosts.
    no such identity: /home/bpratt/pit/scenarios/sbo_openshift/venv/sbo_openshift/keys/xyz: No such file or directory
    cloud-user@xxxx: Permission denied (publickey,gssapi-keyex,gssapi-with-mic).
  unreachable: true

PLAY RECAP ***********************************************************************************************************************************
xxxx                 : ok=1    changed=0    unreachable=1    failed=0    skipped=0    rescued=0    ignored=0

```

**Comment by JunqiZhang0** [comment]
Hi @jbpratt , could you pass me your failing playbook?

**Comment by jbpratt** [comment]
@JunqiZhang0 I'll work on writing a reproducer as it is tied up in our automation and configuration heavily. The gist of this failure was "ssh_private_key_file was missing"

**Comment by JunqiZhang0** [comment]
@jbpratt Thank you very much! A reproducer will help a lot for locating the bug, as I tested some failed playbook and teflo successfully outputted the errors

**Comment by JunqiZhang0** [comment]
@jbpratt any update on the reproducer?

**Comment by jbpratt** [comment]
@JunqiZhang0 I was able to work up a reproducer of sorts, but I think it is exposing another issue. I can't find the complete ansible log at all for this, so I can't see what the error message is (I think it is due to the `ansible_ssh_private_key_file` in provision being set to a key that doesn't exist).

This isn't the exact same behavior as before, but it should get you going in the right direction. The difference from the above is that I can successfully run the playbook standalone which has me a bit confused.

```
---
name: reproducer
description: >
  ...

provision:
  - name: "reproducer"
    groups:
      - "test"
    provider:
      name: openstack
      credential: "{{ OPENSTACK_CREDENTIAL }}"
      image: "RHEL-8.3.0-x86_64-production-latest"
      flavor: "ci.m1.small"
      networks:
        - "provider_net_cci_8"
      keypair: "interop"
    ansible_params:
      ansible_user: cloud-user
      ansible_ssh_private_key_file: "non-existent-key"

orchestrate:
  - name: Configure test driver
    orchestrator: ansible
    hosts: test
    ansible_playbook:
      name: setup_driver.yml
```
```
---
- name: Initialize test driver
  hosts: test
  tasks:
    - name: Gather facts
      setup:
```

```
❯ teflo run -t validate -t provision -t orchestrate -s scenario.yml
--------------------------------------------------
Teflo Framework v1.2.1
Copyright (C) 2021, Red Hat, Inc.
--------------------------------------------------
2021-07-14 05:41:57,156 WARNING Scenario workspace was not set, therefore the workspace is automatically assigned to the current working directory. You may experience problems if files needed by teflo do not exists in the scenario workspace.
2021-07-14 05:41:57,291 INFO

2021-07-14 05:41:57,292 INFO                                TEFLO RUN (START)
2021-07-14 05:41:57,292 INFO -------------------------------------------------------------------------------
2021-07-14 05:41:57,292 INFO  * Data Folder           : .teflo/qqazoy58fr
2021-07-14 05:41:57,293 INFO  * Workspace             : /tmp/d
2021-07-14 05:41:57,293 INFO  * Log Level             : info
2021-07-14 05:41:57,294 INFO  * Tasks                 : ['validate', 'provision', 'orchestrate']
2021-07-14 05:41:57,294 INFO  * Scenario              : reproducer
2021-07-14 05:41:57,294 INFO -------------------------------------------------------------------------------

2021-07-14 05:41:57,295 INFO  * Task    : validate
2021-07-14 05:41:57,295 INFO Sending out any notifications that are registered.
2021-07-14 05:41:57,296 INFO ..................................................
2021-07-14 05:41:57,296 INFO Starting tasks on pipeline: notify
2021-07-14 05:41:57,296 WARNING ... no tasks to be executed ...
2021-07-14 05:41:57,297 INFO ..................................................
2021-07-14 05:41:57,298 INFO Starting tasks on pipeline: validate
2021-07-14 05:41:57,298 INFO --> Blaster v0.4.0 <--
2021-07-14 05:41:57,299 INFO Task Execution: Sequential
2021-07-14 05:41:57,299 INFO Tasks:
2021-07-14 05:41:57,300 INFO 1. Task     : reproducer
                                Class    : <class 'teflo.tasks.validate.ValidateTask'>
                                Methods  : ['run']
2021-07-14 05:41:57,300 INFO 2. Task     : reproducer-bpratt
                                Class    : <class 'teflo.tasks.validate.ValidateTask'>
                                Methods  : ['run']
2021-07-14 05:41:57,301 INFO 3. Task     : Configure test driver
                                Class    : <class 'teflo.tasks.validate.ValidateTask'>
                                Methods  : ['run']
2021-07-14 05:41:57,301 INFO ** BLASTER BEGIN **
2021-07-14 05:41:57,302 INFO Validating <class 'teflo.resources.scenario.Scenario'> (reproducer)
2021-07-14 05:41:57,347 INFO Validating <class 'teflo.resources.assets.Asset'> (reproducer-bpratt)
2021-07-14 05:41:57,362 INFO successfully validated scenario Asset against the schema!
2021-07-14 05:41:57,363 INFO Validating <class 'teflo.resources.actions.Action'> (Configure test driver)
2021-07-14 05:41:57,384 INFO successfully validated scenario Action Configure test driver against the schema!
2021-07-14 05:41:57,384 INFO ** BLASTER COMPLETE **
2021-07-14 05:41:57,385 INFO     -> TOTAL DURATION: 0h:0m:0s
2021-07-14 05:41:57,385 INFO ..................................................
2021-07-14 05:41:57,385 INFO  * Task    : provision
2021-07-14 05:41:57,386 INFO Sending out any notifications that are registered.
2021-07-14 05:41:57,386 INFO ..................................................
2021-07-14 05:41:57,387 INFO Starting tasks on pipeline: notify
2021-07-14 05:41:57,387 WARNING ... no tasks to be executed ...
2021-07-14 05:41:57,388 INFO ..................................................
2021-07-14 05:41:57,388 INFO Starting tasks on pipeline: provision
2021-07-14 05:41:57,389 INFO --> Blaster v0.4.0 <--
2021-07-14 05:41:57,389 INFO Task Execution: Concurrent
2021-07-14 05:41:57,392 INFO Tasks:
2021-07-14 05:41:57,392 INFO 1. Task     : reproducer-bpratt
                                Class    : <class 'teflo.tasks.provision.ProvisionTask'>
                                Methods  : ['run']
2021-07-14 05:41:57,393 INFO ** BLASTER BEGIN **
2021-07-14 05:41:57,404 INFO    provisioning asset reproducer-bpratt
2021-07-14 05:41:57,405 INFO Provisioning asset reproducer-bpratt.
2021-07-14 05:41:57,407 INFO Provisioning machines from <class 'teflo.provisioners.ext.os_libcloud_plugin.openstack_libcloud_plugin.OpenstackLibCloudProvisionerPlugin'>
2021-07-14 05:41:57,409 INFO Provisioning node reproducerbpratt_zeuj5.
2021-07-14 05:42:18,609 INFO Successfully booted node reproducerbpratt_zeuj5.
2021-07-14 05:42:18,614 INFO Wait for node reproducerbpratt_zeuj5 to finish building.
2021-07-14 05:42:20,951 INFO 1. VM reproducerbpratt_zeuj5, STATE=pending, rechecking in 20 seconds.
2021-07-14 05:42:43,485 INFO 1. VM reproducerbpratt_zeuj5, STATE=pending, rechecking in 20 seconds.
2021-07-14 05:43:06,474 INFO 1. VM reproducerbpratt_zeuj5, STATE=running
2021-07-14 05:43:06,478 INFO VM reproducerbpratt_zeuj5 successfully finished building!
2021-07-14 05:43:06,481 INFO Node reproducerbpratt_zeuj5 successfully finished building.
2021-07-14 05:43:06,484 WARNING Node reproducerbpratt_zeuj5 does not require fip.
2021-07-14 05:43:06,487 INFO Successfully provisioned node reproducerbpratt_zeuj5.
2021-07-14 05:43:10,149 INFO Successfully provisioned asset reproducer-bpratt with asset_id f67fda95-03ee-4143-a000-0bc098efa0ee.
2021-07-14 05:43:10,161 INFO ** BLASTER COMPLETE **
2021-07-14 05:43:10,164 INFO     -> TOTAL DURATION: 0h:1m:12s
2021-07-14 05:43:10,245 INFO Populating master inventory file with host(s) reproducer-bpratt
2021-07-14 05:43:10,248 INFO ..................................................
2021-07-14 05:43:10,249 INFO  * Task    : orchestrate
2021-07-14 05:43:10,249 INFO Sending out any notifications that are registered.
2021-07-14 05:43:10,250 INFO ..................................................
2021-07-14 05:43:10,251 INFO Starting tasks on pipeline: notify
2021-07-14 05:43:10,252 WARNING ... no tasks to be executed ...
2021-07-14 05:43:10,253 INFO ..................................................
2021-07-14 05:43:10,253 INFO Starting tasks on pipeline: orchestrate
2021-07-14 05:43:10,254 INFO --> Blaster v0.4.0 <--
2021-07-14 05:43:10,254 INFO Task Execution: Sequential
2021-07-14 05:43:10,255 INFO Tasks:
2021-07-14 05:43:10,255 INFO 1. Task     : Configure test driver
                                Class    : <class 'teflo.tasks.orchestrate.OrchestrateTask'>
                                Methods  : ['run']
2021-07-14 05:43:10,256 INFO ** BLASTER BEGIN **
2021-07-14 05:43:10,258 INFO    running orchestration Configure test driver for ['test']
2021-07-14 05:43:10,258 INFO Executing playbook:
2021-07-14 05:43:10,258 INFO Ansible options used: {}
2021-07-14 05:43:10,258 INFO Executing playbook : setup_driver.yml
[WARNING]: Invalid characters were found in group names but not replaced, use -vvvv to see details
2021-07-14 05:43:10,280 INFO 10.0.103.0
2021-07-14 05:43:10,281 ERROR Orchestration failed :
2021-07-14 05:43:10,351 ERROR Failed to run orchestration Configure test driver
2021-07-14 05:43:10,352 ERROR Orchestration failed : Failed to perform  Configure test driver
2021-07-14 05:43:10,353 ERROR A exception was raised while processing task: Configure test driver method: run
Traceback (most recent call last):
  File "/tmp/d/venv/lib64/python3.9/site-packages/blaster/blast.py", line 83, in run
    value = getattr(task_obj, method)()
  File "/tmp/d/venv/lib64/python3.9/site-packages/teflo/tasks/orchestrate.py", line 59, in run
    self.orchestrator.run()
  File "/tmp/d/venv/lib64/python3.9/site-packages/teflo/orchestrators/action_orchestrator.py", line 68, in run
    raise TefloOrchestratorError("Orchestration failed : Failed to perform  %s" % self.plugin.action_name)
teflo.exceptions.TefloOrchestratorError: Orchestration failed : Failed to perform  Configure test driver
2021-07-14 05:43:11,356 INFO ** BLASTER COMPLETE **
2021-07-14 05:43:11,358 INFO     -> TOTAL DURATION: 0h:0m:1s
2021-07-14 05:43:11,360 ERROR One or more tasks got a status of non zero.
2021-07-14 05:43:11,362 INFO Sending out any notifications that are registered.
2021-07-14 05:43:11,366 INFO ..................................................
2021-07-14 05:43:11,368 INFO Starting tasks on pipeline: notify
2021-07-14 05:43:11,369 WARNING ... no tasks to be executed ...
2021-07-14 05:43:11,398 INFO

2021-07-14 05:43:11,398 INFO                                SCENARIO RUN (END)
2021-07-14 05:43:11,399 INFO -------------------------------------------------------------------------------
2021-07-14 05:43:11,399 INFO  * Duration                       : 0h:1m:14s
2021-07-14 05:43:11,400 INFO  * Passed Tasks                   : ['validate', 'provision']
2021-07-14 05:43:11,400 INFO  * Failed Tasks                   : ['orchestrate']
2021-07-14 05:43:11,401 INFO  * Results Folder                 : .teflo/.results
2021-07-14 05:43:11,401 INFO  * Included Scenario Definition   : []
2021-07-14 05:43:11,402 INFO  * Final Scenario Definition      : .teflo/.results/results.yml
2021-07-14 05:43:11,402 INFO -------------------------------------------------------------------------------
2021-07-14 05:43:11,403 INFO TEFLO RUN (RESULT=FAILED)
```

**Comment by JunqiZhang0** [comment]
@jbpratt I think it's failing because you are using ssh key that doesn't exist, I tested the playbook in your orchestrate phase with static host, it run without any problem. Also I tested an invalid playbook for orchestrate phase, teflo printed out the error message without problem. If you have other issue/problem, you are welcome to open a new issue and I will close this.

**Comment by jbpratt** [comment]
@JunqiZhang0 I stated in the post that this was due to a missing ssh key (`ansible_ssh_private_key_file: "non-existent-key"`, this was explicit). Teflo is still not presenting the error, that is not viewable from the output I posted. This is still a bug in teflo.

**Comment by JunqiZhang0** [comment]
@jbpratt looking at it, will get back to you soon

**Comment by JunqiZhang0** [comment]
I'll be out for the coming sprint, @rujutashinde will take a look

**Comment by rujutashinde** [comment]
@jbpratt In this specific scenario, we saw that only specific exceptions were looked at during ssh to the hosts, the one which were getting missed were not  getting reported back, which is what you saw. (e.g in this case KeyImportError) , I have added a check to catch generic exception. I have the change in my branch https://github.com/rujutashinde/teflo/tree/bug_54 , if you would like to test it out.
